packages/flutter/lib/src/services/text_input.dart

```diff
@@ -769,18 +769,23 @@ class TextEditingValue {
 
   /// Creates an instance of this class from a JSON object.
   factory TextEditingValue.fromJSON(Map<String, dynamic> encoded) {
+    final String text = encoded['text'] as String;
+    final TextSelection selection = TextSelection(
+      baseOffset: encoded['selectionBase'] as int? ?? -1,
+      extentOffset: encoded['selectionExtent'] as int? ?? -1,
+      affinity: _toTextAffinity(encoded['selectionAffinity'] as String?) ?? TextAffinity.downstream,
+      isDirectional: encoded['selectionIsDirectional'] as bool? ?? false,
+    );
+    final TextRange composing = TextRange(
+      start: encoded['composingBase'] as int? ?? -1,
+      end: encoded['composingExtent'] as int? ?? -1,
+    );
+    assert(_verifyRange(selection, text));
```

Comment:
We should probably cast this to TextRange before passing to _verifyRange here and in other places e.g.:

```
    assert(_verifyRange(TextRange(selection.start, selection.end), text));
```

Otherwise the ```if (range == TextRange.empty)``` condition will return false as we'll be in the TextSelection equality operator:

```
  @override
  bool operator ==(Object other) {
    if (identical(this, other)) {
      return true;
    }
    if (other is! TextSelection) {
      return false;
    }
    if (!isValid) {
      return !other.isValid;
    }
    return other.baseOffset == baseOffset
        && other.extentOffset == extentOffset
        && (!isCollapsed || other.affinity == affinity)
        && other.isDirectional == isDirectional;
```

Reply:
You're right! I fixed it by checking start and end individually in `_verifyRange` so that callers will never use it wrong.